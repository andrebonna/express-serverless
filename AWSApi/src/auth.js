'use strict'

const generatePolicy = function(principalId, effect, resource) {
    return {
        principalId: principalId,
        policyDocument: {
            Version: '2012-10-17',
            Statement: [{
                Action: 'execute-api:Invoke',
                Effect: effect,
                Resource: resource
            }]
        }
    };
};

module.exports.authorizerFunc = (event, context, callback) => {

    const [, token] = event.authorizationToken.split(' ');
    const [user] = new Buffer(token, 'base64').toString().split(':');

    let permission = 'Allow';
    if (user === 'testDeny') {
        permission = 'Deny';
    }
    //TODO: Make any validation in any Third Party Authorization Provider
    callback(null, generatePolicy(user, permission, event.methodArn));

};