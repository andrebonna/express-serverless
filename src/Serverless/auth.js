const generatePolicy = function(principalId, effect, resource) {
    return {
        principalId: principalId,
        policyDocument: {
            Version: '2012-10-17',
            Statement: [{
                Action: 'execute-api:Invoke',
                Effect: effect,
                Resource: resource
            }]
        }
    };
};

export function authorizerFunc(event, context, callback) {

    let permission = 'Allow';
    if (!event.authorizationToken) {
        permission = 'Deny';
        callback(null, generatePolicy('anonymous', permission, event.methodArn));
    }
    else {
        const [, token] = event.authorizationToken.split(' ');
        const [user] = new Buffer(token, 'base64').toString().split(':');
    
        if (user === 'testDeny') {
            permission = 'Deny';
        }
        //TODO: Make any validation in any Third Party Authorization Provider
        callback(null, generatePolicy(user, permission, event.methodArn));
    } 
}