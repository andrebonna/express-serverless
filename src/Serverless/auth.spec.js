import { authorizerFunc } from './auth';

describe('Auth Lambda', ()=>{
    it('Test Auth Allow', (done) => {
        authorizerFunc({
            authorizationToken: `Basic ${new Buffer('user:password').toString('base64')}`,
            methodArn: 'methodArnMock'
        }, null, (_, response)=>{
            expect(response).toMatchSnapshot();
            done();
        });
    });

    it('Test Auth Deny without token', (done) => {
        authorizerFunc({
            methodArn: 'methodArnMock'
        }, null, (_, response)=>{
            expect(response).toMatchSnapshot();
            done();
        });
    });

    it('Test Auth Deny', (done) => {
        authorizerFunc({
            authorizationToken: `Basic ${new Buffer('testDeny:password').toString('base64')}`,
            methodArn: 'methodArnMock'
        }, null, (_, response)=>{
            expect(response).toMatchSnapshot();
            done();
        });
    });
});
