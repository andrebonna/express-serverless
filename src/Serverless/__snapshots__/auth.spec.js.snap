exports[`Auth Lambda Test Auth Allow 1`] = `
Object {
  "policyDocument": Object {
    "Statement": Array [
      Object {
        "Action": "execute-api:Invoke",
        "Effect": "Allow",
        "Resource": "methodArnMock",
      },
    ],
    "Version": "2012-10-17",
  },
  "principalId": "user",
}
`;

exports[`Auth Lambda Test Auth Deny 1`] = `
Object {
  "policyDocument": Object {
    "Statement": Array [
      Object {
        "Action": "execute-api:Invoke",
        "Effect": "Deny",
        "Resource": "methodArnMock",
      },
    ],
    "Version": "2012-10-17",
  },
  "principalId": "testDeny",
}
`;
